extensions/ql-vscode/src/run-remote-query.ts

```diff
@@ -1,20 +1,51 @@
-import { Uri } from 'vscode';
+import { Uri, window } from 'vscode';
 import * as yaml from 'js-yaml';
 import * as fs from 'fs-extra';
-import { showAndLogErrorMessage, showAndLogInformationMessage } from './helpers';
+import { getOnDiskWorkspaceFolders, showAndLogErrorMessage, showAndLogInformationMessage } from './helpers';
 import { Credentials } from './authentication';
+import * as cli from './cli';
+import { logger } from './logging';
 
 interface Config {
   repositories: string[];
   ref?: string;
-  language: string;
+  language?: string;
 }
 
 // Test "controller" repository and workflow.
 const OWNER = 'dsp-testing';
 const REPO = 'qc-controller';
 
-export default async function runRemoteQuery(credentials: Credentials, uri?: Uri) {
+/**
+ * Finds the language that a query targets.
+ * If it can't be autodetected, prompt the user to specify the language manually.
+ */
+export async function findLanguage(
+  cliServer: cli.CodeQLCliServer,
+  queryUri: Uri | undefined
+): Promise<string> {
+  const uri = queryUri || window.activeTextEditor?.document.uri;
+  if (uri !== undefined) {
+    try {
+      const queryInfo = await cliServer.resolveQueryByLanguage(getOnDiskWorkspaceFolders(), uri);
+      return (Object.keys(queryInfo.byLanguage))[0];
```

Comment:
Before you return this value, I suggest saving it in a `const` variable and logging a message. That way when looking at the extension log, we can immediately tell whether we followed this code path and successfully detected a language.